docs/proposals/volumes.md

```diff
@@ -0,0 +1,428 @@
+| `gitRepo`               | true                          |
+| `secret`                | true                          |
+| `downwardAPI`           | true                          |
+| `gcePersistentDisk`     | false                         |
+| `awsElasticBlockStore`  | false                         |
+| `nfs`                   | false                         |
+| `iscsi`                 | false                         |
+| `glusterfs`             | false                         |
+| `persistentVolumeClaim` | depends on underlying builder |
+| `rbd`                   | false                         |
+| `cinder`                | false                         |
+| `cephfs`                | false                         |
+
+### Kubelet changes
+
+The Kubelet should be modified to perform ownership and label management when required for a volume.
+
+For ownership management the criteria are:
+
+1.  The `pod.Spec.SecurityContext.SupplementalGroups` field is populated
+2.  The volume type supports ownership management
+
+Logic should be added to the `mountExternalVolumes` method that runs a local `chgrp` and `chmod` if
+the pod-level supplemental group is set and the volume supports ownership management:
+
+```go
+package kubelet
+
+type ChgrpRunner interface {
+    Chgrp(path string, gid int) error
+}
+
+type ChmodRunner interface {
+    Chmod(path string, mode os.FileMode) error
+}
+
+type Kubelet struct {
+    chgrpRunner ChgrpRunner
+    chmodRunner ChmodRunner
+
+    ownershipManagementEnabled bool
+}
+
+func (kl *Kubelet) mountExternalVolumes(pod *api.Pod) (kubecontainer.VolumeMap, error) {
+    podSupplementalGroup = pod.PodSecurityContext.PodSupplementalGroup()
+    podSupplementalGroupSet := false
+    if podSupplementalGroup != "" {
+        podSupplementalGroupSet = true
+    }
+
+    podVolumes := make(kubecontainer.VolumeMap)
+
+    for i := range pod.Spec.Volumes {
+        volSpec := &pod.Spec.Volumes[i]
+
+        rootContext, err := kl.getRootDirContext()
+        if err != nil {
+            return nil, err
+        }
+
+        // Try to use a plugin for this volume.
+        internal := volume.NewSpecFromVolume(volSpec)
+        builder, err := kl.newVolumeBuilderFromPlugins(internal, pod, volume.VolumeOptions{RootContext: rootContext}, kl.mounter)
+        if err != nil {
+            glog.Errorf("Could not create volume builder for pod %s: %v", pod.UID, err)
+            return nil, err
+        }
+        if builder == nil {
+            return nil, errUnsupportedVolumeType
+        }
+        err = builder.SetUp()
+        if err != nil {
+            return nil, err
+        }
+
+        if builder.SupportsOwnershipManagement() &&
+           podSupplementalGroupSet {
+            err = kl.chgrpRunner.Chgrp(builder.GetPath(), podSupplementalGroup)
+            if err != nil {
+                return nil, err
+            }
+
+            err = kl.chmodRunner.Chmod(builder.GetPath(), os.FileMode(0770))
```

Comment:
you need to chmod 02770 to the directory and any child dirs - the sticky bit is sort of important.